base/mmap.jl

```diff
@@ -35,134 +25,181 @@ function mmap_grow(len::Integer, prot::Integer, flags::Integer, fd::Integer, off
+
+const INVALID_HANDLE_VALUE = -1
+gethandle(io::SharedMemSpec) = INVALID_HANDLE_VALUE
+function gethandle(io::IO)
+    handle = Base._get_osfhandle(RawFD(fd(io))).handle
+    systemerror("could not get handle for file to map: $(Base.FormatMessage())", handle == -1)
+    return Int(handle)
 end
 
-### Windows implementation ###
+settings(sh::SharedMemSpec) = utf16(sh.name), sh.readonly, sh.create
+settings(io::IO) = Ptr{Cwchar_t}(C_NULL), isreadonly(io), true
 
-@windows_only type SharedMemSpec
-    name :: AbstractString
-    readonly :: Bool
-    create :: Bool
-end
+# Memory mapped file constants
+const PAGE_READONLY          = UInt32(0x02)
+const PAGE_READWRITE         = UInt32(0x04)
+const PAGE_WRITECOPY         = UInt32(0x08)
 
-@windows_only begin
-# Mmapped-array constructor
-function mmap_array{T,N}(::Type{T}, dims::NTuple{N,Integer}, s::Union(IO,SharedMemSpec), offset::FileOffset)
-    if isa(s,IO)
-        hdl = _get_osfhandle(RawFD(fd(s))).handle
-        if Int(hdl) == -1
-            error("could not get handle for file to map: $(FormatMessage())")
-        end
-        name = Ptr{Cwchar_t}(C_NULL)
-        ro = isreadonly(s)
-        create = true
-    else
-        # shared memory
-        hdl = -1
-        name = utf16(s.name)
-        ro = s.readonly
-        create = s.create
-    end
-    len = prod(dims)*sizeof(T)
-    const granularity::Int = ccall(:jl_getallocationgranularity, Clong, ())
-    if len < 0
-        throw(ArgumentError("requested size must be ≥ 0, got $len"))
-    end
-    if len > typemax(Int)-granularity
-        throw(ArgumentError("file is too large ot memory-map on this platform"))
-    end
-    # Set the offset to a page boundary
-    offset_page::FileOffset = div(offset, granularity)*granularity
-    szfile = convert(Csize_t, len + offset)
-    szarray = szfile - convert(Csize_t, offset_page)
-    access = ro ? 4 : 2
-    if create
-        flprotect = ro ? 0x02 : 0x04
-        mmaphandle = ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cint, Cint, Cint, Cwstring),
-            hdl, C_NULL, flprotect, szfile>>32, szfile&typemax(UInt32), name)
-    else
-        mmaphandle = ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cint, Cint, Cwstring),
-            access, true, name)
-    end
-    if mmaphandle == C_NULL
-        error("could not create file mapping: $(FormatMessage())")
+const PAGE_EXECUTE_READ      = UInt32(0x20)
+const PAGE_EXECUTE_READWRITE = UInt32(0x40)
+const PAGE_EXECUTE_WRITECOPY = UInt32(0x80)
+
+const FILE_MAP_COPY          = UInt32(0x01)
+const FILE_MAP_WRITE         = UInt32(0x02)
+const FILE_MAP_READ          = UInt32(0x04)
+const FILE_MAP_EXECUTE       = UInt32(0x20)
+end # @windows_only
+
+# core impelementation of mmap
+type Stream <: IO
+    ptr::Ptr{Void}    # pointer to mmapped-memory
+    handle::Ptr{Void} # only needed on windows for file mapping object
+    len::Int          # amount of memory mapped
+    offset::Int
```

Comment:
If someone on 32bit wants access to the range 10-12GB in a large file, does making this an `Int` mean they are out of luck? Make it a FileOffset instead?